File: v2/src/main/db/store.ts
import Database from 'better-sqlite3';
import { mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import { migrations } from './migrations.js';

export type Db = Database.Database;

export function openStore(filePath: string): Db {
  mkdirSync(dirname(filePath), { recursive: true });

  const db = new Database(filePath);
  db.pragma('journal_mode = WAL');
  db.pragma('foreign_keys = ON');
  db.pragma('synchronous = NORMAL');

  db.exec(`
    CREATE TABLE IF NOT EXISTS schema_version (
      version    INTEGER PRIMARY KEY,
      applied_at INTEGER NOT NULL
    );
  `);

  applyMigrations(db);
  return db;
}

function applyMigrations(db: Db): void {
  const row = db.prepare('SELECT MAX(version) AS v FROM schema_version').get() as
    | { v: number | null }
    | undefined;
  const current = row?.v ?? 0;

  const pending = migrations.filter((m) => m.version > current);
  if (pending.length === 0) return;

  const insertVersion = db.prepare(
    'INSERT INTO schema_version (version, applied_at) VALUES (?, ?)',
  );

  const apply = db.transaction(() => {
    for (const m of pending) {
      db.exec(m.sql);
      insertVersion.run(m.version, Date.now());
    }
  });
  apply();
}

File: v3/src/main/kotlin/io/mex/data/Store.kt
package io.mex.data

import java.nio.file.Files
import java.nio.file.Path
import java.sql.Connection
import java.sql.DriverManager
import java.sql.Statement

class Store private constructor(val conn: Connection) : AutoCloseable {
    init {
        bootstrap()
        applyMigrations()
    }

    private fun bootstrap() {
        exec("PRAGMA journal_mode=WAL")
        exec("PRAGMA foreign_keys=ON")
        exec("PRAGMA synchronous=NORMAL")
        exec(
            """
            CREATE TABLE IF NOT EXISTS schema_version (
              version    INTEGER PRIMARY KEY,
              applied_at INTEGER NOT NULL
            )
            """.trimIndent(),
        )
    }

    private fun applyMigrations() {
        val current = currentVersion()
        val pending = MIGRATIONS.filter { it.version > current }
        if (pending.isEmpty()) return

        conn.autoCommit = false
        try {
            for (m in pending) {
                conn.createStatement().use { st ->
                    for (chunk in m.sql.split(";").map { it.trim() }.filter { it.isNotEmpty() }) {
                        st.executeUpdate(chunk)
                    }
                }
                conn.prepareStatement("INSERT INTO schema_version (version, applied_at) VALUES (?, ?)")
                    .use { ps ->
                        ps.setInt(1, m.version)
                        ps.setLong(2, System.currentTimeMillis())
                        ps.executeUpdate()
                    }
            }
            conn.commit()
        } catch (e: Exception) {
            conn.rollback()
            throw e
        } finally {
            conn.autoCommit = true
        }
    }

    private fun currentVersion(): Int =
        conn.createStatement().use { st ->
            st.executeQuery("SELECT COALESCE(MAX(version), 0) FROM schema_version").use { rs ->
                if (rs.next()) rs.getInt(1) else 0
            }
        }

    private fun exec(sql: String) {
        conn.createStatement().use { it.execute(sql) }
    }

    override fun close() = conn.close()

    companion object {
        fun open(path: Path): Store {
            Files.createDirectories(path.parent)
            // Class.forName not needed in modern JDBC, but be explicit for jpackage builds.
            Class.forName("org.sqlite.JDBC")
            val conn = DriverManager.getConnection("jdbc:sqlite:${path.toAbsolutePath()}")
            conn.createStatement().use(Statement::closeOnCompletion)
            return Store(conn)
        }
    }
}
